### ISSUE TITLE
تشخیص تشکیل امولسیون‌های مضر (Emulsion Detection)

### ISSUE THREAD

**Issue by deblinux123** [Ai-ithub/iFDC---FDDDC#11]
هدف: استفاده از داده‌های زمانی (سنسورها در طول زمان) برای پیش‌بینی تشکیل امولسیون نامطلوب بین نفت و آب.

🔢 ورودی‌ها (ویژگی‌های زمانی):

viscosity, temperature, shear_rate, salinity, pH

🎯 خروجی پیشنهادی:

ستون جدید مثل emulsion_risk → برچسب‌گذاری شده به‌صورت ۰ و ۱
(این ستون باید مانند fluid_loss_risk به‌صورت مصنوعی ساخته شود)

🔧 الگوریتم‌های پیشنهادی:

مدل‌های سری زمانی مثل LSTM یا GRU به همراه مکانیزم Attention

مدل‌های anomaly detection برای کشف رفتار غیرعادی

**Comment by Mehrnoush504** [comment]
کد تولید داده تغییر داده شده این مورد اصلاح خواهد شد
`import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import classification_report, confusion_matrix
from torch.utils.data import DataLoader, TensorDataset
from tqdm import tqdm
import torch.optim as optim

# Set random seeds for reproducibility
torch.manual_seed(42)
np.random.seed(42)
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

## 1. Load and Prepare Dataset
def load_and_prepare_data(file_path, window_size=10):
    """
    Load synthetic drilling data and create emulsion risk label

    Args:
        file_path: Path to CSV file
        window_size: Size of time windows for LSTM

    Returns:
        X, y: Prepared features and label
        feature_scaler: Fitted scaler for features
    """
    # Load data
    data = pd.read_csv(file_path)

    # Create emulsion risk label based on domain knowledge
    # data['emulsion_risk'] = (
    #     (data['viscosity'] > 38) &
    #     (data['shear_rate'] < 85) &
    #     (data['temperature'] < 170) &
    #     (data['ph'].between(6.5, 7.5))
    # ).astype(int)

    # Encode categorical feature (rock_type)
    # le = LabelEncoder()
    # data['rock_type'] = le.fit_transform(data['rock_type'])

    # Select relevant features
    features = ['viscosity', 'temperature', 'shear_rate', 'salinity', 'ph', 'emulsion_risk']

    # Normalize features
    scaler = StandardScaler()
    scaled_features = scaler.fit_transform(data[features])

    # Create sliding windows
    X, y = [], []
    for i in range(len(data) - window_size):
        X.append(scaled_features[i:i+window_size])
        y.append(data['emulsion_risk'].iloc[i+window_size])

    return np.array(X), np.array(y), scaler

## 2. Model Architecture (LSTM with Attention)
class EncoderLSTM(nn.Module):
    def __init__(self, input_dim, hidden_dim, n_layers):
        super(EncoderLSTM, self).__init__()
        self.rnn = nn.LSTM(input_dim, hidden_dim, n_layers,
                          batch_first=True, dropout=0.2)

    def forward(self, src):
        outputs, (hidden, cell) = self.rnn(src)
        return outputs, hidden, cell

class Attention(nn.Module):
    def __init__(self, hidden_dim):
        super(Attention, self).__init__()
        self.attn = nn.Linear(hidden_dim * 2, hidden_dim)
        self.v = nn.Linear(hidden_dim, 1, bias=False)

    def forward(self, encoder_outputs, hidden):
        # hidden shape: (n_layers, batch_size, hidden_dim)
        hidden = hidden[-1].unsqueeze(1)  # Use last layer

        # Repeat hidden for each time step
        hidden = hidden.repeat(1, encoder_outputs.size(1), 1)

        # Calculate attention energies
        energy = torch.tanh(self.attn(torch.cat((hidden, encoder_outputs), dim=2)))
        attention = self.v(energy).squeeze(2)

        return F.softmax(attention, dim=1)

class DecoderLSTMWithAttention(nn.Module):
    def __init__(self, hidden_dim, n_layers):
        super(DecoderLSTMWithAttention, self).__init__()
        self.rnn = nn.LSTM(hidden_dim, hidden_dim, n_layers,
                          batch_first=True, dropout=0.2)
        self.attention = Attention(hidden_dim)
        self.out = nn.Linear(hidden_dim, 1)

    def forward(self, encoder_outputs, hidden, cell):
        # Calculate attention weights
        attn_weights = self.attention(encoder_outputs, hidden)

        # Calculate context vector
        context = attn_weights.unsqueeze(1).bmm(encoder_outputs)

        # Decoder RNN
        output, (hidden, cell) = self.rnn(context, (hidden, cell))

        # Final prediction
        prediction = torch.sigmoid(self.out(output.squeeze(1)))

        return prediction, hidden, cell

class EmulsionPredictor(nn.Module):
    def __init__(self, input_dim, hidden_dim=128, n_layers=2):
        super(EmulsionPredictor, self).__init__()
        self.encoder = EncoderLSTM(input_dim, hidden_dim, n_layers)
        self.decoder = DecoderLSTMWithAttention(hidden_dim, n_layers)

    def forward(self, src):
        # Encoder pass
        encoder_outputs, hidden, cell = self.encoder(src)

        # Decoder pass
        prediction, _, _ = self.decoder(encoder_outputs, hidden, cell)

        return prediction

## 3. Training and Evaluation
def train_model(model, train_loader, val_loader, n_epochs=5):
    criterion = nn.BCELoss()
    optimizer = optim.Adam(model.parameters(), lr=0.001)

    best_val_loss = float('inf')
    patience = 5
    patience_counter = 0

    for epoch in range(n_epochs):
        model.train()
        train_loss = 0

        for batch_x, batch_y in tqdm(train_loader, desc=f"Epoch {epoch+1}"):
            optimizer.zero_grad()
            outputs = model(batch_x)
            loss = criterion(outputs, batch_y)
            loss.backward()
            optimizer.step()
            train_loss += loss.item()

        # Validation
        val_loss = 0
        model.eval()
        with torch.no_grad():
            for batch_x, batch_y in val_loader:
                outputs = model(batch_x)
                val_loss += criterion(outputs, batch_y).item()

        avg_train_loss = train_loss / len(train_loader)
        avg_val_loss = val_loss / len(val_loader)

        print(f"Train Loss: {avg_train_loss:.4f}, Val Loss: {avg_val_loss:.4f}")

        # Early stopping
        if avg_val_loss < best_val_loss:
            best_val_loss = avg_val_loss
            patience_counter = 0
            torch.save(model.state_dict(), 'best_emulsion_model.pth')
        else:
            patience_counter += 1
            if patience_counter >= patience:
                print("Early stopping triggered")
                break

def evaluate_model(model, test_loader):
    model.eval()
    y_true, y_pred = [], []

    with torch.no_grad():
        for batch_x, batch_y in test_loader:
            outputs = model(batch_x)
            y_true.extend(batch_y.cpu().numpy())
            y_pred.extend((outputs > 0.5).float().cpu().numpy())

    print("\nClassification Report:")
    print(classification_report(y_true, y_pred))

    print("\nConfusion Matrix:")
    print(confusion_matrix(y_true, y_pred))

## 4. Main Execution
if __name__ == "__main__":
    # Load and prepare data
    X, y, scaler = load_and_prepare_data("dataset/synthetic_drilling_data.csv")

    # Convert to tensors
    X = torch.FloatTensor(X).to(device)
    y = torch.FloatTensor(y).unsqueeze(1).to(device)

    # Train/val/test split (80/10/10)
    train_idx = int(0.8 * len(X))
    val_idx = int(0.9 * len(X))

    X_train, y_train = X[:train_idx], y[:train_idx]
    X_val, y_val = X[train_idx:val_idx], y[train_idx:val_idx]
    X_test, y_test = X[val_idx:], y[val_idx:]

    # Create DataLoaders
    train_dataset = TensorDataset(X_train, y_train)
    val_dataset = TensorDataset(X_val, y_val)
    test_dataset = TensorDataset(X_test, y_test)

    train_loader = DataLoader(train_dataset, batch_size=128, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=128)
    test_loader = DataLoader(test_dataset, batch_size=128)

    # Initialize model
    model = EmulsionPredictor(input_dim=X_train.shape[2]).to(device)

    # Train model
    train_model(model, train_loader, val_loader)

    # Evaluate on test set
    model.load_state_dict(torch.load('best_emulsion_model.pth'))
    evaluate_model(model, test_loader)`
